pkg/generator/rps_test.go

```diff
@@ -0,0 +1,376 @@
+				1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000,
+				1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000, 1_000_000,
+			},
+			expectedCount: []int{60, 60},
+		},
+		{
+			testName:           "2min_0.5rps",
+			experimentDuration: 2,
+			rpsTarget:          0.5,
+			expectedIAT: []float64{
+				// minute 1
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				// minute 2
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+				2_000_000, 2_000_000, 2_000_000, 2_000_000, 2_000_000,
+			},
+			expectedCount: []int{30, 30},
+		},
+		{
+			testName:           "2min_0.125rps",
+			experimentDuration: 2,
+			rpsTarget:          0.125,
+			expectedIAT: []float64{
+				// minute 1
+				8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000,
+				// minute 2
+				8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000, 8_000_000,
+			},
+			expectedCount: []int{8, 7},
+		},
+	}
+
+	epsilon := 0.01
+
+	for _, test := range tests {
+		t.Run("warm_start"+test.testName, func(t *testing.T) {
+			matrix, minuteCount := GenerateWarmStartFunction(test.experimentDuration, test.rpsTarget)
+
+			if len(matrix) != len(test.expectedIAT) {
+				t.Errorf("Unexpected IAT array size - got: %d, expected: %d", len(matrix), len(test.expectedIAT))
+			}
+			if len(minuteCount) != len(test.expectedCount) {
+				t.Errorf("Unexpected count array size - got: %d, expected: %d", len(minuteCount), len(test.expectedCount))
+			}
+
+			sum := 0.0
+			count := 0
+			currentMinute := 0
+
+			for i := 0; i < len(matrix); i++ {
+				if math.Abs(matrix[i]-test.expectedIAT[i]) > epsilon {
+					t.Error("Unexpected IAT value.")
+				}
+
+				sum += matrix[i]
+				count++
+
+				if int(sum/60_000_000) != currentMinute {
+					if count != test.expectedCount[currentMinute] {
+						t.Error("Unexpected count array value.")
+					}
+
+					currentMinute = int(sum / 60_000_000)
+					count = 0
+				}
+			}
+		})
+	}
+}
+
+func TestColdStartMatrix(t *testing.T) {
+	tests := []struct {
+		testName           string
+		experimentDuration int
+		rpsTarget          float64
+		cooldownSeconds    int
+		expectedIAT        []common.IATArray
+		expectedCount      [][]int
+	}{
+		{
+			testName:           "2min_1rps",
+			experimentDuration: 2,
+			rpsTarget:          1,
+			cooldownSeconds:    10,
```

Comment:
We should add the test for the cooldown period of > 60s. Same for RPS target of <1/60 RPS.

Reply:
I don't think this makes sense for less than 6 seconds since the minimum stable window period in Knative is 6s. 

This parameter is just there to determine the number of functions needed to deploy to generate a cold start trace of requested RPS, primarily because of Knative control plane issues. A higher value of this parameter is needed for higher cold start RPS values in Knative.

Reply:
Ohhh. I understand. See changes https://github.com/vhive-serverless/invitro/pull/557/commits/c15a64330d53903695c6b6439d8c767fb735043d.

I added a warm start test with RPS=0.01 (`6min_0.01rps`) -- invoke a function every 100s, which is even lower RPS than what you suggested (1/60). I picked 100s because of math rounding.

For cold stars, I added functions:
- with 10s cooldown and 0.01 RPS (`6min_0.01rps_10s_cooldown`). 
- with 120s cooldown and 0.0.1 RPS (`6min_0.01rps_120s_cooldown`) - loader figures out it should create two functions here.